src/client/datascience/jupyter/kernels/kernel.ts

```diff
@@ -500,14 +502,18 @@ export class Kernel implements IKernel {
             this.hookedNotebookForEvents.add(notebook);
             notebook.session.kernelSocket.subscribe(this._kernelSocket);
             notebook.session.onDidDispose(() => {
-                traceInfo(
+                traceInfoIfCI(
                     `Kernel got disposed as a result of notebook.onDisposed ${(
                         this.resourceUri || this.notebookDocument.uri
                     ).toString()}`
                 );
-                // this.kernelExecution.cancel();
                 // Ignore when notebook is disposed as a result of failed restarts.
                 if (!this._ignoreNotebookDisposedErrors) {
+                    traceInfo(
```

Comment:
This was crucial, was difficult to tell why a kernel was disposed.
@rchiodo @IanMatthewHuff FYI - When closing a notebook, VSC un-selects a controller, hence kernels get disposed here.
I wasn't expecting that at all, I was hoping the notebook closed event would get triggered first, but thats not true.